File: sdks/typescript/src/__tests__/client.test.ts
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { BuilderEngine } from '../client';
import {
  AuthenticationError,
  RateLimitError,
  ValidationError,
  NotFoundError,
  BuilderEngineError,
} from '../exceptions';

describe('BuilderEngine Client', () => {
  let client: BuilderEngine;
  const mockFetch = vi.fn();

  beforeEach(() => {
    vi.stubGlobal('fetch', mockFetch);
    client = new BuilderEngine({ apiKey: 'test-api-key' });
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  describe('constructor', () => {
    it('should create client with default options', () => {
      const c = new BuilderEngine({ apiKey: 'test-key' });
      expect(c).toBeInstanceOf(BuilderEngine);
    });

    it('should create client with custom options', () => {
      const c = new BuilderEngine({
        apiKey: 'test-key',
        baseUrl: 'https://custom.api.com',
        timeout: 60000,
        maxRetries: 5,
      });
      expect(c).toBeInstanceOf(BuilderEngine);
    });

    it('should throw error without API key', () => {
      expect(() => new BuilderEngine({ apiKey: '' })).toThrow('API key is required');
    });
  });

  describe('request', () => {
    it('should make successful GET request', async () => {
      const mockResponse = { id: 'agent_123', name: 'Test Agent' };
      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: () => Promise.resolve(mockResponse),
        headers: new Headers(),
      });

      const result = await client.request({
        method: 'GET',
        path: '/api/v1/agents/agent_123',
      });

      expect(result).toEqual(mockResponse);
      expect(mockFetch).toHaveBeenCalledWith(
        'https://api.builderengine.io/api/v1/agents/agent_123',
        expect.objectContaining({
          method: 'GET',
          headers: expect.objectContaining({
            'Authorization': 'Bearer test-api-key',
            'Content-Type': 'application/json',
          }),
        })
      );
    });

    it('should make successful POST request with body', async () => {
      const mockResponse = { id: 'agent_new', name: 'New Agent' };
      const requestBody = { name: 'New Agent', voice: 'nova' };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 201,
        json: () => Promise.resolve(mockResponse),
        headers: new Headers(),
      });

      const result = await client.request({
        method: 'POST',
        path: '/api/v1/agents',
        body: requestBody,
      });

      expect(result).toEqual(mockResponse);
      expect(mockFetch).toHaveBeenCalledWith(
        'https://api.builderengine.io/api/v1/agents',
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify(requestBody),
        })
      );
    });

    it('should include query params in GET request', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: () => Promise.resolve({ data: [] }),
        headers: new Headers(),
      });

      await client.request({
        method: 'GET',
        path: '/api/v1/agents',
        params: { limit: 10, offset: 0, status: 'active' },
      });

      expect(mockFetch).toHaveBeenCalledWith(
        'https://api.builderengine.io/api/v1/agents?limit=10&offset=0&status=active',
        expect.any(Object)
      );
    });

    it('should handle 401 authentication error', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 401,
        json: () => Promise.resolve({ error: { message: 'Invalid API key' } }),
        headers: new Headers(),
      });

      await expect(
        client.request({ method: 'GET', path: '/api/v1/agents' })
      ).rejects.toThrow(AuthenticationError);
    });

    it('should handle 429 rate limit error', async () => {
      const headers = new Headers();
      headers.set('Retry-After', '30');

      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 429,
        json: () => Promise.resolve({ error: { message: 'Rate limit exceeded' } }),
        headers,
      });

      try {
        await client.request({ method: 'GET', path: '/api/v1/agents' });
        expect.fail('Should have thrown');
      } catch (error) {
        expect(error).toBeInstanceOf(RateLimitError);
        expect((error as RateLimitError).retryAfter).toBe(30);
      }
    });

    it('should handle 422 validation error', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 422,
        json: () =>
          Promise.resolve({
            error: {
              message: 'Validation failed',
              details: [{ field: 'name', message: 'Name is required' }],
            },
          }),
        headers: new Headers(),
      });

      try {
        await client.request({ method: 'POST', path: '/api/v1/agents', body: {} });
        expect.fail('Should have thrown');
      } catch (error) {
        expect(error).toBeInstanceOf(ValidationError);
        expect((error as ValidationError).errors).toBeDefined();
      }
    });

    it('should handle 404 not found error', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 404,
        json: () => Promise.resolve({ error: { message: 'Agent not found' } }),
        headers: new Headers(),
      });

      await expect(
        client.request({ method: 'GET', path: '/api/v1/agents/nonexistent' })
      ).rejects.toThrow(NotFoundError);
    });

    it('should retry on server errors', async () => {
      mockFetch
        .mockResolvedValueOnce({
          ok: false,
          status: 500,
          json: () => Promise.resolve({ error: { message: 'Internal error' } }),
          headers: new Headers(),
        })
        .mockResolvedValueOnce({
          ok: true,
          status: 200,
          json: () => Promise.resolve({ id: 'agent_123' }),
          headers: new Headers(),
        });

      const result = await client.request({ method: 'GET', path: '/api/v1/agents/agent_123' });

      expect(result).toEqual({ id: 'agent_123' });
      expect(mockFetch).toHaveBeenCalledTimes(2);
    });

    it('should give up after max retries', async () => {
      mockFetch.mockResolvedValue({
        ok: false,
        status: 500,
        json: () => Promise.resolve({ error: { message: 'Internal error' } }),
        headers: new Headers(),
      });

      await expect(
        client.request({ method: 'GET', path: '/api/v1/agents/agent_123' })
      ).rejects.toThrow(BuilderEngineError);

      expect(mockFetch).toHaveBeenCalledTimes(4); // Initial + 3 retries
    });
  });

  describe('resources', () => {
    it('should have agents resource', () => {
      expect(client.agents).toBeDefined();
    });

    it('should have calls resource', () => {
      expect(client.calls).toBeDefined();
    });

    it('should have conversations resource', () => {
      expect(client.conversations).toBeDefined();
    });

    it('should have phoneNumbers resource', () => {
      expect(client.phoneNumbers).toBeDefined();
    });

    it('should have voices resource', () => {
      expect(client.voices).toBeDefined();
    });

    it('should have webhooks resource', () => {
      expect(client.webhooks).toBeDefined();
    });

    it('should have knowledgeBase resource', () => {
      expect(client.knowledgeBase).toBeDefined();
    });

    it('should have workflows resource', () => {
      expect(client.workflows).toBeDefined();
    });

    it('should have campaigns resource', () => {
      expect(client.campaigns).toBeDefined();
    });

    it('should have analytics resource', () => {
      expect(client.analytics).toBeDefined();
    });

    it('should have organizations resource', () => {
      expect(client.organizations).toBeDefined();
    });

    it('should have users resource', () => {
      expect(client.users).toBeDefined();
    });

    it('should have apiKeys resource', () => {
      expect(client.apiKeys).toBeDefined();
    });

    it('should have billing resource', () => {
      expect(client.billing).toBeDefined();
    });
  });

  describe('streaming', () => {
    it('should return streaming client', () => {
      const streaming = client.streaming();
      expect(streaming).toBeDefined();
      expect(typeof streaming.on).toBe('function');
      expect(typeof streaming.connect).toBe('function');
    });
  });
});

describe('Agents Resource', () => {
  let client: BuilderEngine;
  const mockFetch = vi.fn();

  beforeEach(() => {
    vi.stubGlobal('fetch', mockFetch);
    client = new BuilderEngine({ apiKey: 'test-api-key' });
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('should list agents', async () => {
    const mockResponse = {
      data: [{ id: 'agent_1' }, { id: 'agent_2' }],
      total: 2,
      limit: 10,
      offset: 0,
    };
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 200,
      json: () => Promise.resolve(mockResponse),
      headers: new Headers(),
    });

    const result = await client.agents.list({ limit: 10 });

    expect(result.data).toHaveLength(2);
    expect(result.total).toBe(2);
  });

  it('should create agent', async () => {
    const mockAgent = { id: 'agent_new', name: 'Test', voice: 'nova' };
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 201,
      json: () => Promise.resolve(mockAgent),
      headers: new Headers(),
    });

    const result = await client.agents.create({
      name: 'Test',
      voice: 'nova',
      systemPrompt: 'You are helpful.',
    });

    expect(result.id).toBe('agent_new');
  });

  it('should get agent by id', async () => {
    const mockAgent = { id: 'agent_123', name: 'Test' };
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 200,
      json: () => Promise.resolve(mockAgent),
      headers: new Headers(),
    });

    const result = await client.agents.get('agent_123');

    expect(result.id).toBe('agent_123');
  });

  it('should update agent', async () => {
    const mockAgent = { id: 'agent_123', name: 'Updated' };
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 200,
      json: () => Promise.resolve(mockAgent),
      headers: new Headers(),
    });

    const result = await client.agents.update('agent_123', { name: 'Updated' });

    expect(result.name).toBe('Updated');
  });

  it('should delete agent', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 204,
      json: () => Promise.resolve(undefined),
      headers: new Headers(),
    });

    await expect(client.agents.delete('agent_123')).resolves.not.toThrow();
  });
});

describe('Calls Resource', () => {
  let client: BuilderEngine;
  const mockFetch = vi.fn();

  beforeEach(() => {
    vi.stubGlobal('fetch', mockFetch);
    client = new BuilderEngine({ apiKey: 'test-api-key' });
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('should create outbound call', async () => {
    const mockCall = { id: 'call_123', status: 'queued' };
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 201,
      json: () => Promise.resolve(mockCall),
      headers: new Headers(),
    });

    const result = await client.calls.create({
      agentId: 'agent_123',
      toNumber: '+14155551234',
    });

    expect(result.id).toBe('call_123');
    expect(result.status).toBe('queued');
  });

  it('should get call status', async () => {
    const mockCall = { id: 'call_123', status: 'in_progress', duration: 45 };
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 200,
      json: () => Promise.resolve(mockCall),
      headers: new Headers(),
    });

    const result = await client.calls.get('call_123');

    expect(result.status).toBe('in_progress');
    expect(result.duration).toBe(45);
  });

  it('should end call', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 200,
      json: () => Promise.resolve({ success: true }),
      headers: new Headers(),
    });

    await expect(client.calls.end('call_123')).resolves.not.toThrow();
  });

  it('should get call transcript', async () => {
    const mockTranscript = {
      callId: 'call_123',
      messages: [
        { role: 'assistant', content: 'Hello!' },
        { role: 'user', content: 'Hi there' },
      ],
    };
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 200,
      json: () => Promise.resolve(mockTranscript),
      headers: new Headers(),
    });

    const result = await client.calls.getTranscript('call_123');

    expect(result.messages).toHaveLength(2);
  });

  it('should get call recording', async () => {
    const mockRecording = { url: 'https://cdn.example.com/recording.mp3' };
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 200,
      json: () => Promise.resolve(mockRecording),
      headers: new Headers(),
    });

    const result = await client.calls.getRecording('call_123');

    expect(result.url).toContain('recording.mp3');
  });
});

describe('Campaigns Resource', () => {
  let client: BuilderEngine;
  const mockFetch = vi.fn();

  beforeEach(() => {
    vi.stubGlobal('fetch', mockFetch);
    client = new BuilderEngine({ apiKey: 'test-api-key' });
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('should create campaign', async () => {
    const mockCampaign = { id: 'camp_123', name: 'Test Campaign', status: 'draft' };
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 201,
      json: () => Promise.resolve(mockCampaign),
      headers: new Headers(),
    });

    const result = await client.campaigns.create({
      name: 'Test Campaign',
      agentId: 'agent_123',
    });

    expect(result.id).toBe('camp_123');
    expect(result.status).toBe('draft');
  });

  it('should start campaign', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 200,
      json: () => Promise.resolve({ success: true }),
      headers: new Headers(),
    });

    await expect(client.campaigns.start('camp_123')).resolves.not.toThrow();
  });

  it('should pause campaign', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 200,
      json: () => Promise.resolve({ success: true }),
      headers: new Headers(),
    });

    await expect(client.campaigns.pause('camp_123')).resolves.not.toThrow();
  });

  it('should get campaign progress', async () => {
    const mockProgress = {
      total: 100,
      completed: 45,
      failed: 5,
      pending: 50,
    };
    mockFetch.mockResolvedValueOnce({
      ok: true,
      status: 200,
      json: () => Promise.resolve(mockProgress),
      headers: new Headers(),
    });

    const result = await client.campaigns.getProgress('camp_123');

    expect(result.total).toBe(100);
    expect(result.completed).toBe(45);
  });
});

File: sdks/go/builderengine/client_test.go
package builderengine

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestNewClient(t *testing.T) {
	t.Run("creates client with API key", func(t *testing.T) {
		client, err := NewClient("test-api-key")
		require.NoError(t, err)
		assert.NotNil(t, client)
	})

	t.Run("returns error without API key", func(t *testing.T) {
		client, err := NewClient("")
		assert.Error(t, err)
		assert.Nil(t, client)
	})

	t.Run("applies custom options", func(t *testing.T) {
		client, err := NewClient("test-api-key",
			WithBaseURL("https://custom.api.com"),
			WithTimeout(60*time.Second),
			WithMaxRetries(5),
		)
		require.NoError(t, err)
		assert.NotNil(t, client)
	})
}

func TestClientResources(t *testing.T) {
	client, err := NewClient("test-api-key")
	require.NoError(t, err)

	t.Run("has agents resource", func(t *testing.T) {
		assert.NotNil(t, client.Agents)
	})

	t.Run("has calls resource", func(t *testing.T) {
		assert.NotNil(t, client.Calls)
	})

	t.Run("has conversations resource", func(t *testing.T) {
		assert.NotNil(t, client.Conversations)
	})

	t.Run("has phoneNumbers resource", func(t *testing.T) {
		assert.NotNil(t, client.PhoneNumbers)
	})

	t.Run("has voices resource", func(t *testing.T) {
		assert.NotNil(t, client.Voices)
	})

	t.Run("has webhooks resource", func(t *testing.T) {
		assert.NotNil(t, client.Webhooks)
	})

	t.Run("has knowledgeBase resource", func(t *testing.T) {
		assert.NotNil(t, client.KnowledgeBase)
	})

	t.Run("has workflows resource", func(t *testing.T) {
		assert.NotNil(t, client.Workflows)
	})

	t.Run("has campaigns resource", func(t *testing.T) {
		assert.NotNil(t, client.Campaigns)
	})

	t.Run("has analytics resource", func(t *testing.T) {
		assert.NotNil(t, client.Analytics)
	})

	t.Run("has organizations resource", func(t *testing.T) {
		assert.NotNil(t, client.Organizations)
	})

	t.Run("has users resource", func(t *testing.T) {
		assert.NotNil(t, client.Users)
	})

	t.Run("has apiKeys resource", func(t *testing.T) {
		assert.NotNil(t, client.APIKeys)
	})

	t.Run("has billing resource", func(t *testing.T) {
		assert.NotNil(t, client.Billing)
	})
}

func TestClientRequest(t *testing.T) {
	t.Run("makes successful GET request", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "GET", r.Method)
			assert.Equal(t, "Bearer test-api-key", r.Header.Get("Authorization"))
			assert.Equal(t, "application/json", r.Header.Get("Content-Type"))

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"id":   "agent_123",
				"name": "Test Agent",
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		var result map[string]interface{}
		err := client.Request(context.Background(), RequestOptions{
			Method: "GET",
			Path:   "/api/v1/agents/agent_123",
		}, &result)

		require.NoError(t, err)
		assert.Equal(t, "agent_123", result["id"])
	})

	t.Run("makes successful POST request", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "POST", r.Method)

			var body map[string]interface{}
			json.NewDecoder(r.Body).Decode(&body)
			assert.Equal(t, "New Agent", body["name"])

			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusCreated)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"id":   "agent_new",
				"name": "New Agent",
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		var result map[string]interface{}
		err := client.Request(context.Background(), RequestOptions{
			Method: "POST",
			Path:   "/api/v1/agents",
			Body:   map[string]string{"name": "New Agent"},
		}, &result)

		require.NoError(t, err)
		assert.Equal(t, "agent_new", result["id"])
	})

	t.Run("includes query params", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "10", r.URL.Query().Get("limit"))
			assert.Equal(t, "active", r.URL.Query().Get("status"))

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{"data": []interface{}{}})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		var result map[string]interface{}
		err := client.Request(context.Background(), RequestOptions{
			Method: "GET",
			Path:   "/api/v1/agents",
			Params: map[string]string{"limit": "10", "status": "active"},
		}, &result)

		require.NoError(t, err)
	})
}

func TestErrorHandling(t *testing.T) {
	t.Run("handles 401 authentication error", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusUnauthorized)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"error": map[string]interface{}{
					"message": "Invalid API key",
					"code":    "invalid_api_key",
				},
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		var result map[string]interface{}
		err := client.Request(context.Background(), RequestOptions{
			Method: "GET",
			Path:   "/api/v1/agents",
		}, &result)

		require.Error(t, err)
		assert.True(t, IsAuthenticationError(err))
	})

	t.Run("handles 404 not found error", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusNotFound)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"error": map[string]interface{}{
					"message": "Agent not found",
				},
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		var result map[string]interface{}
		err := client.Request(context.Background(), RequestOptions{
			Method: "GET",
			Path:   "/api/v1/agents/nonexistent",
		}, &result)

		require.Error(t, err)
		assert.True(t, IsNotFoundError(err))
	})

	t.Run("handles 422 validation error", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusUnprocessableEntity)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"error": map[string]interface{}{
					"message": "Validation failed",
					"details": map[string]interface{}{
						"name": "Name is required",
					},
				},
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		var result map[string]interface{}
		err := client.Request(context.Background(), RequestOptions{
			Method: "POST",
			Path:   "/api/v1/agents",
			Body:   map[string]string{},
		}, &result)

		require.Error(t, err)
		assert.True(t, IsValidationError(err))
	})

	t.Run("handles 429 rate limit error", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			w.Header().Set("Retry-After", "30")
			w.WriteHeader(http.StatusTooManyRequests)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"error": map[string]interface{}{
					"message": "Rate limit exceeded",
				},
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL), WithMaxRetries(0))

		var result map[string]interface{}
		err := client.Request(context.Background(), RequestOptions{
			Method: "GET",
			Path:   "/api/v1/agents",
		}, &result)

		require.Error(t, err)
		assert.True(t, IsRateLimitError(err))
		if rlErr, ok := err.(*RateLimitError); ok {
			assert.Equal(t, 30, rlErr.RetryAfter)
		}
	})
}

func TestAgentsService(t *testing.T) {
	t.Run("lists agents", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"data":   []map[string]interface{}{{"id": "agent_1"}, {"id": "agent_2"}},
				"total":  2,
				"limit":  10,
				"offset": 0,
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		result, err := client.Agents.List(context.Background(), AgentListParams{})

		require.NoError(t, err)
		assert.Len(t, result.Data, 2)
		assert.Equal(t, 2, result.Total)
	})

	t.Run("creates agent", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			var body map[string]interface{}
			json.NewDecoder(r.Body).Decode(&body)
			assert.Equal(t, "Test Agent", body["name"])

			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusCreated)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"id":   "agent_new",
				"name": "Test Agent",
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		result, err := client.Agents.Create(context.Background(), AgentCreateParams{
			Name:         "Test Agent",
			Voice:        "nova",
			SystemPrompt: "You are helpful.",
		})

		require.NoError(t, err)
		assert.Equal(t, "agent_new", result.ID)
	})

	t.Run("gets agent", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"id":   "agent_123",
				"name": "Test Agent",
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		result, err := client.Agents.Get(context.Background(), "agent_123")

		require.NoError(t, err)
		assert.Equal(t, "agent_123", result.ID)
	})

	t.Run("deletes agent", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "DELETE", r.Method)
			w.WriteHeader(http.StatusNoContent)
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		err := client.Agents.Delete(context.Background(), "agent_123")

		require.NoError(t, err)
	})
}

func TestCallsService(t *testing.T) {
	t.Run("creates call", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusCreated)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"id":     "call_123",
				"status": "queued",
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		result, err := client.Calls.Create(context.Background(), CallCreateParams{
			AgentID:  "agent_123",
			ToNumber: "+14155551234",
		})

		require.NoError(t, err)
		assert.Equal(t, "call_123", result.ID)
		assert.Equal(t, CallStatus("queued"), result.Status)
	})

	t.Run("ends call", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "POST", r.Method)
			assert.Contains(t, r.URL.Path, "/end")
			w.WriteHeader(http.StatusOK)
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		err := client.Calls.End(context.Background(), "call_123")

		require.NoError(t, err)
	})

	t.Run("gets transcript", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"call_id": "call_123",
				"messages": []map[string]interface{}{
					{"role": "assistant", "content": "Hello!"},
					{"role": "user", "content": "Hi there"},
				},
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		result, err := client.Calls.GetTranscript(context.Background(), "call_123")

		require.NoError(t, err)
		assert.Equal(t, "call_123", result.CallID)
		assert.Len(t, result.Messages, 2)
	})
}

func TestCampaignsService(t *testing.T) {
	t.Run("creates campaign", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusCreated)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"id":     "camp_123",
				"name":   "Test Campaign",
				"status": "draft",
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		result, err := client.Campaigns.Create(context.Background(), CampaignCreateParams{
			Name:    "Test Campaign",
			AgentID: "agent_123",
		})

		require.NoError(t, err)
		assert.Equal(t, "camp_123", result.ID)
		assert.Equal(t, CampaignStatus("draft"), result.Status)
	})

	t.Run("starts campaign", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Contains(t, r.URL.Path, "/start")
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"id":     "camp_123",
				"status": "running",
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		result, err := client.Campaigns.Start(context.Background(), "camp_123")

		require.NoError(t, err)
		assert.Equal(t, CampaignStatus("running"), result.Status)
	})

	t.Run("gets progress", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"total":      100,
				"completed":  45,
				"successful": 40,
				"failed":     5,
				"pending":    55,
			})
		}))
		defer server.Close()

		client, _ := NewClient("test-api-key", WithBaseURL(server.URL))

		result, err := client.Campaigns.GetProgress(context.Background(), "camp_123")

		require.NoError(t, err)
		assert.Equal(t, 100, result.Total)
		assert.Equal(t, 45, result.Completed)
	})
}

func TestStreaming(t *testing.T) {
	t.Run("creates streaming client", func(t *testing.T) {
		client, _ := NewClient("test-api-key")
		streaming := client.Streaming()

		assert.NotNil(t, streaming)
	})
}
